### ISSUE TITLE
Dynamo adapter can't deploy to edge

### ISSUE THREAD

**Issue by arackaf** [nextauthjs/next-auth#6582]
### Environment

I'm using SvelteKit so I ran 

```
npx envinfo --system --binaries --browsers --npmPackages "@sveltejs/kit,svelte,@auth/*"
```

```
  System:
    OS: macOS 13.1
    CPU: (16) x64 Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
    Memory: 35.95 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.11.1 - /usr/local/Cellar/nvm/0.38.0/versions/node/v16.11.1/bin/node
    Yarn: 1.22.17 - /usr/local/Cellar/nvm/0.38.0/versions/node/v16.11.1/bin/yarn
    npm: 9.3.0 - ~/Documents/git/booklist/svelte-kit/node_modules/.bin/npm
  Browsers:
    Chrome: 109.0.5414.119
    Firefox: 109.0.1
```

### Reproduction URL

Not quite a repro, but this shows the error I'm reporting:

https://vercel.com/adam-rackis/booklist-svelte-kit/9KDXU5pMSbXbNsHS3nvk8MKNrqdi

EDIT - repro is here: https://github.com/arackaf/dynamo-adapter-edge-deploy

`npm i` then `npm run build` and there should be an error about `crypto` not existing. See comments below for more info

### Describe the issue

When trying to deploy to edge with Vercel, the dynamo adapter causes things to fail 

>  Error: Build failed with 2 errors:
--
00:37:29.818 | .svelte-kit/output/server/entries/endpoints/api/cover-upload/_server.ts.js:1:17: ERROR: Could not resolve "path"
00:37:29.818 | node_modules/@next-auth/dynamodb-adapter/dist/index.js:4:25: ERROR: Could not resolve "crypto"
00:37:29.818 | at failureErrorWithLog (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1604:15)

### How to reproduce

Use the dynamo adapter, and try to deploy to the edge (with SvelteKit and the vercel adapter, but presumably with Next, too)

### Expected behavior

Ideally it would be great if this worked. Is it possible for the adapter to use the web crypto api, instead of Node's?

**Comment by balazsorban44** [comment]
The Vercel logs aren't public. Please add a proper reproduction

**Comment by arackaf** [comment]
Really sorry - after pulling this into a minimal repro, the problem is my path import, not crypto being used in dynamo-adapter. That's pretty obvious now, looking at it again. But geez the stack trace is easily misread. 



Error: Build failed with 2 errors:
--
00:37:29.818 | .svelte-kit/output/server/entries/endpoints/api/cover-upload/_server.ts.js:1:17: ERROR: Could not resolve "path"
00:37:29.818 | node_modules/@next-auth/dynamodb-adapter/dist/index.js:4:25: ERROR: Could not resolve "crypto"
00:37:29.818 | at failureErrorWithLog (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1604:15)
00:37:29.818 | at /vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1056:28
00:37:29.819 | at /vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1001:67
00:37:29.819 | at buildResponseToResult (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1054:7)
00:37:29.819 | at /vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:1166:14
00:37:29.819 | at responseCallbacks.<computed> (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:701:9)
00:37:29.819 | at handleIncomingPacket (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:756:9)
00:37:29.819 | at Socket.readFromStdout (/vercel/path0/svelte-kit/node_modules/esbuild/lib/main.js:677:7)
00:37:29.819 | at Socket.emit (node:events:513:28)
00:37:29.819 | at Socket.emit (node:domain:489:12) {
00:37:29.820 | errors: [
00:37:29.820 | {
00:37:29.820 | detail: undefined,
00:37:29.820 | id: '',
00:37:29.820 | location: {
00:37:29.820 | column: 17,
00:37:29.820 | file: '.svelte-kit/output/server/entries/endpoints/api/cover-upload/_server.ts.js',
00:37:29.821 | length: 6,
00:37:29.821 | line: 1,
00:37:29.821 | lineText: 'import path from "path";',
00:37:29.821 | namespace: '',
00:37:29.821 | suggestion: ''
00:37:29.821 | },
00:37:29.821 | notes: [
00:37:29.821 | {
00:37:29.822 | location: null,
00:37:29.822 | text: `The package "path" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.`
00:37:29.822 | }
00:37:29.822 | ],
00:37:29.822 | pluginName: '',
00:37:29.822 | text: 'Could not resolve "path"'
00:37:29.822 | },
00:37:29.823 | {
00:37:29.823 | detail: undefined,
00:37:29.823 | id: '',
00:37:29.823 | location: {
00:37:29.823 | column: 25,
00:37:29.823 | file: 'node_modules/@next-auth/dynamodb-adapter/dist/index.js',
00:37:29.823 | length: 8,
00:37:29.823 | line: 4,
00:37:29.824 | lineText: 'const crypto_1 = require("crypto");',
00:37:29.824 | namespace: '',
00:37:29.824 | suggestion: ''
00:37:29.824 | },
00:37:29.824 | notes: [
00:37:29.824 | {
00:37:29.824 | location: null,
00:37:29.824 | text: `The package "crypto" wasn't found on the file system but is built into node. Are you trying to bundle for node? You can use "platform: 'node'" to do that, which will remove this error.`
00:37:29.825 | }
00:37:29.825 | ],
00:37:29.825 | pluginName: '',
00:37:29.825 | text: 'Could not resolve "crypto"'
00:37:29.825 | }
00:37:29.825 | ],
00:37:29.825 | warnings: []
00:37:29.826 | }

**Comment by balazsorban44** [comment]
Agreed! I let the SvelteKit team now. :pray:

**Comment by arackaf** [comment]
@balazsorban44 - no actually this is broken. My path import does break edge deploy. But after removing that, I do get the same `crypto` error.

Here's a minimal repro: 

https://github.com/arackaf/dynamo-adapter-edge-deploy 

clone the project, `npm i` and then `npm run build` and you should see this error:

<img width="948" alt="image" src="https://user-images.githubusercontent.com/11261266/216228659-7d4ebd87-d432-4e63-ae71-3ea583fdf7b7.png">

**Comment by balazsorban44** [comment]
Oh yeah...
https://github.com/nextauthjs/next-auth/blob/326eadf0edad2a49a92545df44623304cc521958/packages/adapter-dynamodb/src/index.ts#L1

So it's not necessarily a bug, we just haven't started working on making adapters fully web standard compatible. 

We can probably create a new major version that ships ESM-only and targets environments where `crypto` is available, or use `uuid` instead thta we do with other adapters too for id generation.

The question is, do you know if `@aws-sdk/client-dynamodb` and `@aws-sdk/lib-dynamodb` are actually compatible? Cause we do depend on those.

**Comment by arackaf** [comment]
@balazsorban44 good question - this would all be for naught of the underlying aws utils aren't compatible. 

Lemme check that now ...

**Comment by balazsorban44** [comment]
Underlying lib https://github.com/aws/aws-sdk-js-v3/blob/2ed17a3aee6b431b9f54275d98b607969a45b3f1/packages/smithy-client/package.json#L28 not promising... 😬

**Comment by arackaf** [comment]
@balazsorban44 these Dynamo libs appear to work fine - this modified hooks file does build successfully. 

I'm not familiar with the smithy-client package - is that used internally to the dynamo client/lib packages?

```js
import { sequence } from '@sveltejs/kit/hooks';

import { SvelteKitAuth } from '@auth/sveltekit';
import { DynamoDB } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocument } from '@aws-sdk/lib-dynamodb';
import GoogleProvider from '@auth/core/providers/google';

const dynamoConfig = {};

export const client = DynamoDBDocument.from(new DynamoDB(dynamoConfig), {
	marshallOptions: {
		convertEmptyValues: true,
		removeUndefinedValues: true,
		convertClassInstanceToMap: true
	}
});

const auth = SvelteKitAuth({
	providers: [
		// @ts-ignore
		GoogleProvider({
			clientId: 'GOOGLE_AUTH_CLIENT_ID',
			clientSecret: 'GOOGLE_AUTH_SECRET'
		})
	],
	session: {
		maxAge: 60 * 60 * 24 * 365,
		strategy: 'jwt'
	},

	secret: 'NEXTAUTH_SECRET'
});

export async function handleFn({ event, resolve }) {
	if (event.url.pathname.startsWith('/custom')) {
		return new Response('custom response');
	}

	const response = await resolve(event);
	return response;
}

export const handle = sequence(handleFn, auth);

```

**Comment by balazsorban44** [comment]
yeah, it seems to be an internal package. but I think there's a good reason to get a new version of this package out anyway, will try to do it soon-ish, and you can test if it will work. 🙏

**Comment by arackaf** [comment]
Eeek! Thank you! Hit me up (here or Twitter DM) and I'll test it asap

**Comment by arackaf** [comment]
@balazsorban44 just fyi, I ran a better test on this. I created a branch, and deployed that separately to Vercel, with the dynamo adapter completely removed.

uuid failed the build, but only because it wants 

```js
import { v4 } from "uuid";
```

instead of 

```js
import uuid from "uuid"
const { v4 } = uuid;
```

After fixing that it built, but the edge function crashed pretty hard

> ReferenceError: global is not defined
>     at (../../../../node_modules/buffer/index.js:43:29)
>     at (../../../../.svelte-kit/vercel-tmp/render/edge.js:12:7)
>     at (../../../../node_modules/aws-sdk/lib/browserHashUtils.js:1:13)
>     at (../../../../.svelte-kit/vercel-tmp/render/edge.js:12:7)
>     at (../../../../node_modules/aws-sdk/lib/browserHmac.js:1:16)
>     at (../../../../.svelte-kit/vercel-tmp/render/edge.js:12:7)
>     at (../../../../node_modules/aws-sdk/lib/browserCryptoLib.js:1:11)
>     at (../../../../.svelte-kit/vercel-tmp/render/edge.js:12:7)
>     at (../../../../node_modules/aws-sdk/lib/browser_loader.js:4:18)
>     at (../../../../.svelte-kit/vercel-tmp/render/edge.js:12:7)

Honestly not sure where `aws-sdk/lib/browserHashUtils` is being pulled in from. That path of my code shouldn't really be doing anything with aws, or even with the browser's Hash

**Comment by arackaf** [comment]
@balazsorban44 also, obv this would also be blocked by https://github.com/nextauthjs/next-auth/issues/6439 but that's a separate issue.

Loving this library. Thank you!

**Comment by ThangHuuVu** [comment]
Hi @arackaf I've just published the experimental packages that made the DynamoDB adapter ESM only and use UUID here: https://www.npmjs.com/package/@next-auth/dynamodb-adapter/v/2.0.0 Could you give it a try? 🙌

**Comment by arackaf** [comment]
@ThangHuuVu fyi I think the problem is here

```js
"main": "dist/index.js",
```

but the install in my node_modules shows no dist folder at all

**Comment by ThangHuuVu** [comment]
Whoops sorry for that - I've re-published v2.0.0-experimental.1 to fix that. When I try to install and run the new package in your reproduction this is the new error:

```
[vite]: Rollup failed to resolve import "@aws-sdk/lib-dynamodb" from "/Users/thangvu/repos/github/dynamo-adapter-edge-deploy/src/hooks.server.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
error during build:
Error: [vite]: Rollup failed to resolve import "@aws-sdk/lib-dynamodb" from "/Users/thangvu/repos/github/dynamo-adapter-edge-deploy/src/hooks.server.js".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
    at onRollupWarning (file:///Users/thangvu/repos/github/dynamo-adapter-edge-deploy/node_modules/vite/dist/node/chunks/dep-3007b26d.js:44772:19)
    at onwarn (file:///Users/thangvu/repos/github/dynamo-adapter-edge-deploy/node_modules/vite/dist/node/chunks/dep-3007b26d.js:44542:13)
    at Object.onwarn (file:///Users/thangvu/repos/github/dynamo-adapter-edge-deploy/node_modules/rollup/dist/es/shared/rollup.js:24537:13)
    at ModuleLoader.handleInvalidResolvedId (file:///Users/thangvu/repos/github/dynamo-adapter-edge-deploy/node_modules/rollup/dist/es/shared/rollup.js:23221:26)
    at file:///Users/thangvu/repos/github/dynamo-adapter-edge-deploy/node_modules/rollup/dist/es/shared/rollup.js:23181:26
```

after adding `"@aws-sdk/lib-dynamodb": "latest"` to `package.json` `dependencies`, I got a successful build

**Comment by arackaf** [comment]
@ThangHuuVu hmmm - not sure what repro of mine you added to. When I upgrade to experimental.1 and run npm run build I get this error 

> Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/arackis/Documents/git/booklist/svelte-kit/node_modules/@next-auth/dynamodb-adapter/utils' imported from /Users/arackis/Documents/git/booklist/svelte-kit/node_modules/@next-auth/dynamodb-adapter/index.js

When I just run dev and try to run, the import 

```js
import { DynamoDBAdapter } from "@next-auth/dynamodb-adapter";
```

is undefined. Looking in the deploy of `@next-auth/dynamodb-adapter` it looks like the `utils` folder wasn't published, but is referenced by the exports field in package.json

**Comment by arackaf** [comment]
Oh you're talking about this repro 

https://github.com/arackaf/dynamo-adapter-edge-deploy 

hmm - when I upgrade to 

> "@next-auth/dynamodb-adapter": "^2.0.0-experimental.1",

I get the same 

> Error [ERR_MODULE_NOT_FOUND]: Cannot find module '/Users/arackis/Documents/git/dynamo-adapter-edge-deploy/node_modules/@next-auth/dynamodb-adapter/utils' imported from /Users/arackis/Documents/git/dynamo-adapter-edge-deploy/node_modules/@next-auth/dynamodb-adapter/index.js
>     at new NodeError (node:internal/errors:371:5)
>     at finalizeResolution (node:internal/modules/esm/resolve:418:11)
>     at moduleResolve (node:internal/modules/esm/resolve:983:10)
>     at defaultResolve (node:internal/modules/esm/resolve:1080:11)
>     at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
>     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
>     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
>     at link (node:internal/modules/esm/module_job:78:36) {
>   code: 'ERR_MODULE_NOT_FOUND'
> }

I did also try installing 

> "@aws-sdk/lib-dynamodb": "^3.264.0",

but that had no effect (it would be surprising if it did - it does appear the utils folder is missing)

**Comment by ThangHuuVu** [comment]
@arackaf Okay sorry, I got no idea why it worked for me in the first place. I've published a new package to fix that, it has to work this time 😄 https://www.npmjs.com/package/@next-auth/dynamodb-adapter/v/2.0.0-experimental.3?activeTab=explore

Could you try it?

**Comment by arackaf** [comment]
@ThangHuuVu sorry I missed this. It works! I've got it running in an edge worker, writing to my dynamo db.

Thank you SO MUCH for the fast turnaround 🫡

I hope this can get deployed to a stable version soon 🙏

**Comment by balazsorban44** [comment]
@arackaf it is out [@next-auth/dynamodb-adapter@3.0.0](https://github.com/nextauthjs/next-auth/releases/tag/%40next-auth%2Fdynamodb-adapter%403.0.0).

(We skipped 2.x due to a publish accident :grimacing:, but should not matter much :pray:)

**Comment by arackaf** [comment]
@balazsorban44 lol hell of a publish accident that an entire major version was wiped away :P 

Hope there's some decent release notes so people don't get too scared by the double-major upgrade

**Comment by balazsorban44** [comment]
Well, we published 2.0.0 by accident and did not revert, and our release bot figured since we are doing breaking changes, it should go to 3.0.0 instead. :shrug:

I can add a note about this in the [release notes](https://github.com/nextauthjs/next-auth/releases/tag/%40next-auth%2Fdynamodb-adapter%403.0.0), thanks!
